add config option to show rule URI in output

Something like:

```
[shellcheck] bin/command.sh:6:6 Double quote to prevent globbing and word splitting. (SC2086) <https://www.shellcheck.net/wiki/SC2086>
```

On by default, but users can disable if it's too busy.

--- a/internal/cmd/shared.go
+++ b/internal/cmd/shared.go
@@ -29,6 +29,9 @@ func addOutputFlags(cmd *cobra.Command, oc *stylist.OutputConfig) {
 	cmd.Flags().BoolVar(
 		&oc.ShowContext, "show-context", oc.ShowContext, "Show the lines of code affected",
 	)
+	cmd.Flags().BoolVar(
+		&oc.ShowURL, "show-url", oc.ShowURL, "Show issue URLs when available",
+	)
 }
 
 func addProcessorFilterFlags(cmd *cobra.Command, filter *stylist.ProcessorFilter) {

--- a/internal/stylist/config.go
+++ b/internal/stylist/config.go
@@ -24,6 +24,7 @@ type Config struct {
 type OutputConfig struct {
 	Format      ResultFormat `yaml:"format,omitempty"       default:"tty"`
 	ShowContext bool         `yaml:"show_context,omitempty" default:"true"`
+	ShowURL     bool         `yaml:"show_url,omitempty" default:"true"`
 }
 
 func NewConfig() *Config {

--- a/internal/stylist/presets.yml
+++ b/internal/stylist/presets.yml
@@ -77,10 +77,20 @@ hadolint:
   includes:
     - "**/Dockerfile"
   check:
-    command: "hadolint --format=sarif --no-color"
+    command: "hadolint --format=json --no-color"
     input: variadic
     output: stdout
-    format: sarif
+    # Using json (vs. sarif) because the sarif output doesn't include the URL.
+    format: json
+    mapping:
+      level: "{{ .level }}"
+      path: "{{ .file }}"
+      start_line: "{{ .line }}"
+      start_column: "{{ .column }}"
+      rule_id: "{{ .code }}"
+      rule_name: "{{ .code }}"
+      rule_description: "{{ .message }}"
+      rule_uri: '{{ if hasPrefix "SC" .code }}https://www.shellcheck.net/wiki/{{ .code }}{{ else }}https://github.com/hadolint/hadolint/wiki/{{ .code }}{{ end }}'
 
 markdownlint:
   name: "markdownlint"
@@ -102,7 +112,7 @@ markdownlint:
       end_column: "{{ if .errorRange }}{{ add (index .errorRange 0) (index .errorRange 1) }}{{ end }}"
       rule_id: "{{ index .ruleNames 0 }}"
       rule_name: "{{ index .ruleNames 1 }}"
-      rule_description: "{{ .ruleDescription }} {{ if .errorDetail }}[{{ .errorDetail }}]{{ end }}"
+      rule_description: "{{ .ruleDescription }}"
       rule_uri: "{{ .ruleInformation }}"
   fix:
     command: "markdownlint --fix"

--- a/internal/stylist/result_printer.go
+++ b/internal/stylist/result_printer.go
@@ -67,12 +67,24 @@ func (p *TtyPrinter) Print(results []*Result) error {
 func (p *TtyPrinter) printLocation(result *Result, formatter *ioutil.Formatter) {
 	msg := ""
 	if result.Rule.Description != "" {
-		msg = formatter.Red(result.Rule.Description)
+		switch result.Level {
+		case ResultLevelError:
+			msg = formatter.Red(result.Rule.Description)
+		case ResultLevelWarning:
+			msg = formatter.Yellow(result.Rule.Description)
+		case ResultLevelNote:
+			msg = formatter.Blue(result.Rule.Description)
+		default:
+			msg = result.Rule.Description
+		}
 	}
 
 	rule := ""
 	if result.Rule.ID != "" {
 		rule = fmt.Sprintf("(%s)", result.Rule.ID)
+	}
+	if result.Rule.URI != "" && p.config.Output.ShowURL {
+		rule = fmt.Sprintf("%s <%s>", rule, result.Rule.URI)
 	}
 
 	fmt.Fprintf(
@@ -98,10 +110,12 @@ func (p *TtyPrinter) printUnderLinePointer(result *Result, formatter *ioutil.For
 		return
 	}
 
-	col0 := result.Location.StartColumn - 1
+	startCol0 := result.Location.StartColumn - 1
+	endCol0 := result.Location.EndColumn - 1
 	line := result.ContextLines[0]
+
 	prefixRunes := make([]rune, 0, len(line))
-	for j := 0; j < len(line) && j < col0; j++ {
+	for j := 0; j < len(line) && j < startCol0; j++ {
 		if line[j] == '\t' {
 			prefixRunes = append(prefixRunes, '\t')
 		} else {
@@ -109,5 +123,14 @@ func (p *TtyPrinter) printUnderLinePointer(result *Result, formatter *ioutil.For
 		}
 	}
 
-	fmt.Fprintf(p.ios.Out, "%s%s\n", string(prefixRunes), formatter.Yellow("^"))
+	indicatorCols := 1
+	if endCol0 > startCol0 && endCol0 <= len(line) {
+		indicatorCols = endCol0 - startCol0
+	}
+	indicatorRunes := make([]rune, 0, indicatorCols)
+	for j := 0; j < indicatorCols; j++ {
+		indicatorRunes = append(indicatorRunes, '^')
+	}
+
+	fmt.Fprintf(p.ios.Out, "%s%s\n", string(prefixRunes), formatter.Yellow(string(indicatorRunes)))
 }

--- a/internal/stylist/result_printer_test.go
+++ b/internal/stylist/result_printer_test.go
@@ -22,3 +22,176 @@ func TestNewResultPrinter(t *testing.T) {
 		_ = NewResultPrinter(ios, config)
 	})
 }
+
+func TestTtyPrinter_Print(t *testing.T) {
+	results := []*Result{
+		{
+			Source: "test-linter",
+			Level:  ResultLevelError,
+			Location: ResultLocation{
+				Path:        "some/path/foo.go",
+				StartLine:   1,
+				StartColumn: 0,
+			},
+			Rule: ResultRule{
+				ID:          "rule-id1",
+				Name:        "rule-name1",
+				Description: "no start column",
+				URI:         "https://test-linter.com/rule-id1",
+			},
+			ContextLines: []string{
+				"context line one",
+			},
+		},
+		{
+			Source: "test-linter",
+			Level:  ResultLevelWarning,
+			Location: ResultLocation{
+				Path:        "some/path/bar.go",
+				StartLine:   2,
+				StartColumn: 10,
+				EndLine:     2,
+				EndColumn:   14,
+			},
+			Rule: ResultRule{
+				ID:          "rule-id2",
+				Name:        "rule-name2",
+				Description: "valid start and end column",
+				URI:         "https://test-linter.com/rule-id2",
+			},
+			ContextLines: []string{
+				"\tcontext line two",
+			},
+		},
+		{
+			Source: "test-linter",
+			Level:  ResultLevelNote,
+			Location: ResultLocation{
+				Path:        "some/path/baz.go",
+				StartLine:   1,
+				StartColumn: 1,
+			},
+			Rule: ResultRule{
+				ID:          "rule-id3",
+				Name:        "rule-name3",
+				Description: "single char indicator",
+				URI:         "https://test-linter.com/rule-id3",
+			},
+			ContextLines: []string{
+				"context line three",
+			},
+		},
+		{
+			Source: "test-linter",
+			Level:  ResultLevelNone,
+			Location: ResultLocation{
+				Path:        "some/path/qux.go",
+				StartLine:   1,
+				StartColumn: 99,
+				EndLine:     1,
+				EndColumn:   999,
+			},
+			Rule: ResultRule{
+				ID:          "rule-id4",
+				Name:        "rule-name4",
+				Description: "out of bounds indicator",
+				URI:         "https://test-linter.com/rule-id4",
+			},
+			ContextLines: []string{
+				"context line four",
+			},
+		},
+	}
+
+	tests := []struct {
+		desc     string
+		config   OutputConfig
+		results  []*Result
+		expected []string
+		err      string
+	}{
+		{
+			desc:     "empty result set should print nothing",
+			results:  []*Result{},
+			expected: []string{},
+			err:      "",
+		},
+
+		{
+			desc: "should print a minimal result set when config disabled",
+			config: OutputConfig{
+				ShowContext: false,
+				ShowURL:     false,
+			},
+			results: results,
+			expected: []string{
+				"[test-linter] some/path/foo.go:1:0 no start column (rule-id1)",
+				"[test-linter] some/path/bar.go:2:10 valid start and end column (rule-id2)",
+				"[test-linter] some/path/baz.go:1:1 single char indicator (rule-id3)",
+				"[test-linter] some/path/qux.go:1:99 out of bounds indicator (rule-id4)",
+			},
+			err: "",
+		},
+		{
+			desc: "should print context when enabled",
+			config: OutputConfig{
+				ShowContext: true,
+				ShowURL:     false,
+			},
+			results: results,
+			expected: []string{
+				"[test-linter] some/path/foo.go:1:0 no start column (rule-id1)",
+				"context line one",
+				"[test-linter] some/path/bar.go:2:10 valid start and end column (rule-id2)",
+				"\tcontext line two",
+				"\t        ^^^^",
+				"[test-linter] some/path/baz.go:1:1 single char indicator (rule-id3)",
+				"context line three",
+				"^",
+				"[test-linter] some/path/qux.go:1:99 out of bounds indicator (rule-id4)",
+				"context line four",
+				"                 ^",
+			},
+			err: "",
+		},
+		{
+			desc: "should print urls when enabled",
+			config: OutputConfig{
+				ShowContext: false,
+				ShowURL:     true,
+			},
+			results: results,
+			expected: []string{
+				"[test-linter] some/path/foo.go:1:0 no start column (rule-id1) " +
+					"<https://test-linter.com/rule-id1>",
+				"[test-linter] some/path/bar.go:2:10 valid start and end column (rule-id2) " +
+					"<https://test-linter.com/rule-id2>",
+				"[test-linter] some/path/baz.go:1:1 single char indicator (rule-id3) " +
+					"<https://test-linter.com/rule-id3>",
+				"[test-linter] some/path/qux.go:1:99 out of bounds indicator (rule-id4) " +
+					"<https://test-linter.com/rule-id4>",
+			},
+			err: "",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			app := NewTestApp()
+			app.Config.Output = tt.config
+
+			printer := &TtyPrinter{
+				ios:    app.IO,
+				config: app.Config,
+			}
+			err := printer.Print(tt.results)
+
+			if tt.err == "" {
+				assert.NoError(t, err)
+			} else {
+				assert.ErrorContains(t, err, tt.err)
+			}
+
+			assert.Equal(t, tt.expected, app.IO.Out.Lines())
+		})
+	}
+}

